### PR TITLE
Issue #15685: fixed JavadocParagraph to detect paragraph tag when they have their corresponding closing tag

### LINKED ISSUE #15685: JavadocParagraph does not work when paragraphs have their corresponding closing tag
I have read check documentation: https://checkstyle.org/checks/javadoc/javadocparagraph.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Detected at: https://github.com/checkstyle/checkstyle/issues/15503#issuecomment-2362869286

In [JavadocParagraph](https://checkstyle.org/checks/javadoc/javadocparagraph.html)'s implementation, we're only checking for `P` tags which aren't closed, i.e we're checking only for `<p>` which doesn't have corresponding `</p>`.

There is a difference in parsed AST when `P` have its corresponding closing tag and when it doesn't have it. As explained in the above referenced comment, AST for `P` when it is not closed looks like this:

```
    |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [4:3]
    |   |   |       |   `--P_TAG_START -> P_TAG_START [4:3]
```

and when it is properly closed, it looks like this:

```
    |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [4:3]
    |   |   |       |   `--PARAGRAPH -> PARAGRAPH [4:3]
    |   |   |       |       |--P_TAG_START -> P_TAG_START [4:3]
```

A new `PARAGRAPH ` appears between the token.

What we do in the Check's implementation:

https://github.com/checkstyle/checkstyle/blob/d07072b179330e40357c7c8852883af815d6978c/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L147-L156

We check if after `HTML_ELEMENT` is there a `P_TAG_START` token? This token is for opening paragraph tag ( `<p>` ). If yes then move further.

This case is only valid for `<p>` tags which does not have their corresponding `</p>`, so when we encounter a proper pair of paragraph tag (`<p></p>`), our check ignores that paragraph as a new token ( `PARAGRAPH` ) appears between `HTML_ELEMENT` & `P_TAG_START`.

Example:

```xml
$ cat .\javadocparagraph_config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocParagraph"/>
  </module>
</module>
```

```java
$ cat .\InvalidParagraphTagPosition.java
/**
 * Some summary.
 *
 * <p>  Some Javadoc.
 *
 * <p>  Some Javadoc. </p> // false-negative
 */
public class InvalidParagraphTagPosition {
    /**
     * some javadoc.
     * <p>
     *
     * <p></p> // false-negative
     */
    int a;

    /**
     * Some Javadoc.<P>
     * Some Javadoc.<P></P> // false-negative
     */
    int b;
}
```

```
$ java -jar .\checkstyle-10.18.1-all.jar -c .\javadocparagraph_config.xml .\InvalidParagraphTagPosition.java
Starting audit...
[ERROR] C:\checkstyle testing\.\InvalidParagraphTagPosition.java:4: <p> tag should be placed immediately before the first word, with no space after. [JavadocParagraph]
[ERROR] C:\checkstyle testing\.\InvalidParagraphTagPosition.java:11: <p> tag should be placed immediately before the first word, with no space after. [JavadocParagraph]
[ERROR] C:\checkstyle testing\.\InvalidParagraphTagPosition.java:11: <p> tag should be preceded with an empty line. [JavadocParagraph]
[ERROR] C:\checkstyle testing\.\InvalidParagraphTagPosition.java:18: <p> tag should be placed immediately before the first word, with no space after. [JavadocParagraph]
[ERROR] C:\checkstyle testing\.\InvalidParagraphTagPosition.java:18: <p> tag should be preceded with an empty line. [JavadocParagraph]
Audit done.
Checkstyle ends with 5 errors.
```

We get errors only for `<p>` which are not closed, we don't get any violations for `<p>` which are closed properly. We expect the same errors as we got for `<p>`

---

**Describe what you expect in detail.**

Fix the check's implementation to check for both types of tags: `<p>` & `<p></p>`

### PR DESCRIPTION
**Pull request by Zopsss**
fixes #15685 https://github.com/checkstyle/checkstyle/issues/15732

tried fixing the implementation of JavadocParagraph ~might have introduced new false-negatives :p~

I tried to make the Check work same way when the paragraph tag's corresponding closing tag is also present.

when I tried to run build command locally, I got so many errors for different files like checks, filters, etc for `<p>` tag. Most of them were for:

```
<p> tag should be placed immediately before the first word, with no space after.
```

@romani need your help here

---

Diff Regression config: https://gist.githubusercontent.com/Zopsss/013f379d4c821f2d624be56212a64575/raw/5f17ec9186d4421396429ab38cbac3e4663bf4eb/javadocparagraph_config.xml
Diff Regression projects: https://raw.githubusercontent.com/checkstyle/contribution/master/checkstyle-tester/projects-to-test-on-for-github-action.properties

### CODE REVIEW

**Comment by romani** [comment]
we need to use for checkstyle-checks.xml 
`<property name="allowNewlineParagraph" value="false"/>`

**Comment by Zopsss** [comment]
I have set the `allowNewlineParagraph` to false in checkstyle-checks.xml and now there is no violation for

```
<p> tag should be placed immediately before the first word, with no space after.
```

though there are still following errors in the build failure:

```
<p> tag should be preceded with an empty line.
Redundant <p> tag.
```

Locally I tried to fix `allowNewlineParagraph` by doing something like this:

```java

    /**
     * Tests whether the paragraph tag is misplaced.
     * Meaning that it is not immediate before the first word or another tag.
     *
     * @param tag html tag.
     * @return true, if the paragraph tag is misplaced.
     */
    private boolean isMisplacedParagraph(DetailNode tag) {
        boolean shouldGiveViolation;

        if (allowNewlineParagraph) {
            shouldGiveViolation = false;
        }
        else if (JavadocUtil.getFirstChild(tag).getType() == JavadocTokenTypes.PARAGRAPH) {
            final DetailNode paragraphToken = JavadocUtil.getFirstChild(tag);
            final DetailNode paragraphStartTagToken = JavadocUtil.getFirstChild(paragraphToken);
            final DetailNode nextSibling = JavadocUtil.getNextSibling(paragraphStartTagToken);
            shouldGiveViolation = nextSibling.getType() == JavadocTokenTypes.NEWLINE
                    || nextSibling.getType() == JavadocTokenTypes.EOF
                    || nextSibling.getText().startsWith(" ");
        }
        else {
            final DetailNode nextSibling = JavadocUtil.getNextSibling(tag);
            shouldGiveViolation = nextSibling.getType() == JavadocTokenTypes.NEWLINE
                    || nextSibling.getType() == JavadocTokenTypes.EOF
                    || nextSibling.getText().startsWith(" ");
        }

        return shouldGiveViolation;
    }
```

```java

    /**
     * Determines whether or not the line with paragraph tag has previous empty line.
     *
     * @param tag html tag.
     */
    private void checkParagraphTag(DetailNode tag) {
        final DetailNode newLine = getNearestEmptyLine(tag);
        if (isFirstParagraph(tag)) {
            log(tag.getLineNumber(), MSG_REDUNDANT_PARAGRAPH);
        }
        else if (newLine == null || tag.getLineNumber() - newLine.getLineNumber() != 1) {
            log(tag.getLineNumber(), MSG_LINE_BEFORE);
        }
        if (isMisplacedParagraph(tag)) { // this line is changed.
            log(tag.getLineNumber(), MSG_MISPLACED_TAG);
        }
    }
```

With this there was no error of:

```
<p> tag should be placed immediately before the first word, with no space after.
```

and I also didn't needed to set `allowNewlineParagraph` to false. There were still plenty of other errors... which are other false-positives of the check.

**Comment by romani** [comment]
this Check code is very hard to read, lets refactor it a bit before this PR to proceeed.

please look at https://github.com/checkstyle/checkstyle/pull/15709 and try to finish it.
We need to split newline handling to make logic a bit more clear.

**Comment by Zopsss** [comment]
build is still failing, there are still lots of errors

**Comment by romani** [comment]
before we proceed on this, all other PRs and issues should be merged.
https://github.com/checkstyle/checkstyle/issues/15713 must be done before, as it will make inputs more realistic so we will make update more clearly.

**Comment by romani** [comment]
@Zopsss , please resolve conflicts in this PR to let me start review.
you can create new Inputs to avoid conflicts with pending PR for Block-tag.

**Comment by Zopsss** [comment]
> we need to use for checkstyle-checks.xml
> `<property name="allowNewlineParagraph" value="false"/>`

reverted this as we fixed this in one of my old PR.

**Comment by Zopsss** [comment]
local build was still failing, it was mostly giving errors for:

```
<p> tag should be preceded with an empty line.
Redundant <p> tag.
```

It was also giving this error:

```
<p> tag should be placed immediately before the first word, with no space after.
```

for cases like:

```
<p> testing: </p>
   ↑ whitespace 
```

the above error is fine, it is violation of rule but other 2 errors are false-positive, they started appearing because we added support to check for open-close tag. 

Should we keep these errors or not? If yes, then it would require changing lots of files.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/37a777c3f69b162e88bb6fc88659fe4595248b29/config/checkstyle-checks.xml#L720

This means that allowNewlineParagraph= true

And it means that you have defect, as this property is to allow \n.

**Comment by Zopsss** [comment]
> And it means that you have defect, as this property is to allow \n.

yes we allow it, so that check does not give violation on lines like:

https://github.com/checkstyle/checkstyle/blob/37a777c3f69b162e88bb6fc88659fe4595248b29/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L28-L31

the default property is what we need for our style. We set it to false for google_checks.xml because it does not follow such rules.

**Comment by romani** [comment]
Google style use false, checkstyle style is true.
We need to improve logic for allowNewlineParagraph to do not violate our javadocs that are keep paragraph tag alone on line.

File: config/checkstyle-checks.xml
Comment (Zopsss) [inline config/checkstyle-checks.xml:60]: suppression until we fix the implementation of this check. May change the referenced issue in future if we open new issue for it

**Comment by Zopsss** [comment]
A PR has been sent for failing CI: https://checkstyle.semaphoreci.com/jobs/46bc3abc-9190-43d9-bb99-3a25a47a386b

https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/1037

**Comment by Zopsss** [comment]
fixed some bug in the latest push: https://github.com/checkstyle/checkstyle/compare/643b58d3d4f9225d764a9a064a5aa5e5eac69503..c80a41ccedf6312677b6578b19091f815bb3d1e1#diff-19c14e011c79f51ad2e551f4941a1def0d5ee70c40d93fd944e07da166bad4bb

**Comment by romani** [comment]
@Zopsss , please send PR to pgjdbc project to fix violations.

**Comment by Zopsss** [comment]
Build is flooded with errors again :(

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/28127/workflows/60b27b2a-fb0d-4774-a64f-b37dea3364ca/jobs/698909

this time it is because of following type of violation:

```
<p> tag should not precede HTML block-tag '<ul>', <p> tag should be removed. [JavadocParagraph]
```

should I add suppression for this too?

**Comment by Zopsss** [comment]
GitHub, generate report for configs in PR description

File: config/checkstyle-checks.xml
Comment (romani) [reply config/checkstyle-checks.xml:60]: We need to fix almost all violations from Paragraph Check. https://github.com/checkstyle/checkstyle/blob/aeaf4b0838d814ce62ef23aa6b4d367ff5f83fb2/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java#L84-L86 is good example. Please start a process of updating our sources, let's make PRs to be limited to ~20  files. All PRs will have "supplemental: " prefix.

Comment (romani) [reply config/checkstyle-checks.xml:60]: If there is some strange cases, report to us and skip them for now, we will handle them at the end of as soon as you get confirmation on how to update javadoc.

**Comment by Zopsss** [comment]
GitHub, generate report for configs in PR description

**Comment by Zopsss** [comment]
Fixed https://github.com/checkstyle/checkstyle/pull/15686#issuecomment-2391546197 in the latest push: https://github.com/checkstyle/checkstyle/compare/1042402e60a306559e7a2f263389b8ec7d91ddaa..de2acad15324993d03afdf25fac93db4dcbe9fc5

**Comment by Zopsss** [comment]
GitHub, generate report for configs in PR description

File: src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java:145]: this is actually false positives.
Google rule for P should be applicable only for P tags  that are not nested in other HTML tags.
in other words they are on root level of javadoc.

`<p> tag should be preceded with an empty line.` this is not html rule, html does not care about empty lines.
this rule is for humans who read javadoc as source, humans more easy to parse javadoc source in this case. I doubt empty line will do a difference in resulted rendered html from javadoc.

`'<p> tag should be placed immediately before the first word'` this is also not html related rule, it is just style of Google.

When first P is open, all nested in it, is HTML, and will have content to render well in browser, there could be any layout, we should not deal with it, it is very complicated and I doubt anybody care about this.

Lets keep our logic simple, if we detect P tag, and it is nested in other tag, we skip it.

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java
Comment (Zopsss) [reply src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java:145]: should we check if it precedes a block-tag or we should skip that too?

Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java:145]: if P nested, no violations at all, no any violations.

**Comment by romani** [comment]
CI failure:
```
[INFO] --- jacoco-maven-plugin:0.8.12:check (default-check) @ checkstyle ---
[INFO] Loading execution data file C:\projects\checkstyle\target\jacoco.exec
[INFO] Analyzed bundle 'checkstyle' with 837 classes
[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle
  .checks.javadoc.JavadocParagraphCheck: branches covered ratio is 0.99, but expected minimum is 1.00
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:49]: this might be  false-negative.
Indentation spaces should trigger a violation.

this is no violation code:
```
    * <p>
    * Some Javadoc.
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:84]: should be violation

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag2.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag2.java:89]: should be violation.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag3.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag3.java:17]: please update documentation that this Check ignores HTML comments.

**Review by romani — CHANGES_REQUESTED** [review]
please update doc to describe new behavior.

items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java
Comment (Zopsss) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:49]: This code was previously suggested by you:

https://github.com/checkstyle/checkstyle/blob/619265de69dabb63264cddaa387fc524c656a0f6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L213-L215

In this input file allowNewlineParagraph is set to true, based on the above code this example should not give violation if allowNewlineParagraph is set to true.

New lines are allowed, so I think this is correct code. `some javadoc` does not need to be right after `<p>`

Comment (Zopsss) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:49]: >* <p>
>    *   Some Javadoc.

If we give violation on this code and use `<p> should be placed immediately before text, without space` message then it would confuse users. Correct me if I'm wrong.

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:49]: Lets not be blocked by this,  we might need to move this discussion on a side to separate issue.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag3.java
Comment (Zopsss) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag3.java:17]: done https://github.com/checkstyle/checkstyle/blob/5e28cccd6ecf9074c5bca15e71ab4b258bd77539/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L44-L46

**Comment by Zopsss** [comment]
GitHub, generate report for configs in PR description

**Comment by Zopsss** [comment]
GitHub, generate report for configs in PR description

**Comment by romani** [comment]
Github, generate report for configs in PR description

**Comment by romani** [comment]
GitHub, generate report for JavadocParagraph/all-examples-in-one

**Comment by Zopsss** [comment]
I tried using the latest commit in validation.sh for pgjdbc

https://checkstyle.semaphoreci.com/jobs/4b4e6720-796f-4bb0-ad18-69345b97de77#L689

but it requires java 17 or 21, current version was 11: https://checkstyle.semaphoreci.com/jobs/4b4e6720-796f-4bb0-ad18-69345b97de77#L718

I looked into validation.sh to update the java version but couldn't find it.

**Comment by romani** [comment]
GitHub, generate report for JavadocParagraph/all-examples-in-one

**Comment by romani** [comment]
@Zopsss , please send PR to pgjdbc  for their latest master
and I am ok to just disable pgjdbc until this PR is merged, We will find way to use jdk17 in semaphore a bit later.

**Comment by romani** [comment]
@Zopsss , please rebase over latest master, I upgraded pgjdbc to SHA that you have in this PR.

**Comment by Zopsss** [comment]
PR sent at pgjdbc https://github.com/pgjdbc/pgjdbc/pull/3397

**Comment by Zopsss** [comment]
@romani finally, all CIs are passing. I have fixed incorrect paragraphs at pgjdbc. Please do the final review.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java:50]: This should be violation, why it is not ?
Even we ignore \n , there are bunch of spaces before UL.
And should be no P before UL, as it is blocktag

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by romani** [comment]
GitHub, generate report for JavadocParagraph/all-examples-in-one

**Comment by romani** [comment]
GitHub, generate report for JavadocParagraph/all-examples-in-one

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java
Comment (Zopsss) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java:51]: > Even we ignore \n , there are bunch of spaces before UL.

we decided not to check this at https://github.com/checkstyle/checkstyle/pull/15686#discussion_r1788518551, read my comment above it for explanation

> And should be no P before UL, as it is blocktag

fixed this, thanks for pointing this out

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java:48]: we need to add something about "if P nested, no violations at all, no any violations."
see https://github.com/checkstyle/checkstyle/pull/15686/files#r1787687046

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java:181]: add 4 spaced before `||` to put it visually under `(` to show priority of execution

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:49]: please create issue

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java:51]: please create issue and put "ok , until #xxxxxxxxx" to let readers know that we detected it but postponed this false-negative

File: src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java
Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java:148]: please document this, users should know abou this.

File: config/checkstyle-checks.xml
Comment (romani) [reply config/checkstyle-checks.xml:60]: we need fixes, please continue https://github.com/checkstyle/checkstyle/pull/15737 and all other fixes.
We eat food we produce before we release to it to users, we need to fix all of them.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java
Comment (Zopsss) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java]: done

https://github.com/checkstyle/checkstyle/blob/cb7cac3b2e167e9fc260dc64576c87ab45257cfd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L48-L53

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java
Comment (Zopsss) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java]: done

https://github.com/checkstyle/checkstyle/blob/cb7cac3b2e167e9fc260dc64576c87ab45257cfd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L184-L188

**Comment by Zopsss** [comment]
not sure why ci is failing

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/28232/workflows/df53342b-c53d-431f-9c77-888fd26fff20/jobs/703338

local build was passing.

**Comment by Zopsss** [comment]
Github, generate site

**Comment by romani** [comment]
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/50771010/job/410wglmmfwmaa7nr#L254

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.12.1:site (default-site) on project checkstyle: Error parsing 'C:\projects\checkstyle\src\xdocs\checks\javadoc\javadocparagraph.xml': Error validating the model: Error:
[ERROR]   Public ID: null
[ERROR]   System ID: null
[ERROR]   Line number: 34
[ERROR]   Column number: 15
[ERROR]   Message: cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://maven.apache.org/XDOC/2.0":ul}'. One of 
```

File: src/xdocs/checks/javadoc/javadocparagraph.xml.template
Comment (romani) [inline src/xdocs/checks/javadoc/javadocparagraph.xml.template:34]: Hmm, this is not allowed tag ... Strange.
Can we workaround it somehow?
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/50771010/job/410wglmmfwmaa7nr#L254

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/xdocs/checks/javadoc/javadocparagraph.xml.template
Comment (romani) [reply src/xdocs/checks/javadoc/javadocparagraph.xml.template:34]: You can reproduce on local by `./.ci/validation.cmd site_without_verify`

Or `mvn site`

**Comment by romani** [comment]
Github, generate site

**Comment by Zopsss** [comment]
GitHub, generate report for JavadocParagraph/all-examples-in-one

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java:51]: Please put new issue in several case of this PR to make it clear it is postponed

File: config/checkstyle-checks.xml
Comment (romani) [reply config/checkstyle-checks.xml:60]: Put this link in comment , you will update all in separate PR

Comment (Zopsss) [reply config/checkstyle-checks.xml:60]: this will be addressed at https://github.com/checkstyle/checkstyle/issues/15762

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java
Comment (Zopsss) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java]: this will be addressed at https://github.com/checkstyle/checkstyle/issues/15762

Comment (Zopsss) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java]: I have added `// ok until 15762` comments

**Review by romani — APPROVED** [review]
Ok to merge

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java:237]: Please rename to `nested`, it's a variable name, shouldn't be a verb.

**Review by rdiachenko — APPROVED** [review]
Minor note, ok to merge after it is fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java
Comment (Zopsss) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java]: done

**Comment by Zopsss** [comment]
Check no closed issue reference failing CI is addressed at https://github.com/checkstyle/checkstyle/pull/15765

**Review by romani — APPROVED** [review]
I see fix for above item.

**Pull request merged**
